File: solutions/cpp/Regular_expression.cpp
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <string>

using std::cout;
using std::endl;
using std::string;

bool IsMatchHere(const string &r, size_t r_idx, const string &s, size_t s_idx);

// @include
bool IsMatch(const string &r, const string &s) {
  // Case (2.): r starts with '^'.
  if (r.front() == '^') {
    return IsMatchHere(r, 1, s, 0);
  }

  for (size_t i = 0; i <= s.size(); ++i) {
    if (IsMatchHere(r, 0, s, i)) {
      return true;
    }
  }
  return false;
}

bool IsMatchHere(const string &r, size_t r_idx, const string &s,
                 size_t s_idx) {
  // Case (1.).
  if (r_idx == r.size()) {
    return true;
  }

  // Case (2): r ends with '$'.
  if (r_idx == r.size() - 1 && r[r_idx] == '$') {
    return s_idx == s.size();
  }

  // Case (4.).
  if (r.size() - r_idx >= 2 && r[r_idx + 1] == '*') {
    // Don't use any character of r[r_idx].
    if (IsMatchHere(r, r_idx + 2, s, s_idx)) {
      return true;
    }

    // Tries all possible number of r[r_idx].
    while (s_idx < s.size() && (r[r_idx] == '.' || r[r_idx] == s[s_idx])) {
      if (IsMatchHere(r, r_idx + 2, s, ++s_idx)) {
        return true;
      }
    }
  }

  // Case (3.).
  return s_idx < s.size() && (r[r_idx] == '.' || r[r_idx] == s[s_idx]) &&
         IsMatchHere(r, r_idx + 1, s, s_idx + 1);
}
// @exclude

int main(int argc, char *argv[]) {
  assert(IsMatch(".", "a"));
  assert(IsMatch("a", "a"));
  assert(!IsMatch("a", "b"));
  assert(IsMatch("a.9", "aW9"));
  assert(!IsMatch("a.9", "aW19"));
  assert(IsMatch("^a.9", "aW9"));
  assert(!IsMatch("^a.9", "baW19"));
  assert(IsMatch(".*", "a"));
  assert(IsMatch(".*", ""));
  assert(IsMatch("c*", "c"));
  assert(!IsMatch("aa*", "c"));
  assert(IsMatch("ca*", "c"));
  assert(IsMatch(".*",  "asdsdsa"));
  assert(IsMatch("9$" , "xxxxW19"));

  assert(IsMatch(".*a", "ba"));

  assert(IsMatch(".*a", "ba"));
  assert(IsMatch("^a.*9$", "axxxxW19"));

  assert(IsMatch("^a.*W19$", "axxxxW19"));
  assert(IsMatch(".*a.*W19", "axxxxW19123"));
  assert(!IsMatch(".*b.*W19", "axxxxW19123"));
  assert(IsMatch("n.*", "n"));
  assert(IsMatch("a*n.*", "an"));
  assert(IsMatch("a*n.*", "ana"));
  assert(IsMatch("a*n.*W19", "anaxxxxW19123"));
  assert(IsMatch(".*a*n.*W19", "asdaaadnanaxxxxW19123"));
  assert(IsMatch(".*.*.a*n.*W19", "asdaaadnanaxxxxW19123"));
  return 0;
}

File: static/cpp/java/src/main/java/com/epi/RegularExpression.java
package com.epi;

public class RegularExpression {
  // @include
  public static boolean isMatch(String r, String s) {
    // Case (2.) : starts with '^'.
    if (r.charAt(0) == '^') {
      return isMatchHere(r.substring(1), s);
    }

    for (int i = 0; i <= s.length(); ++i) {
      if (isMatchHere(r, s.substring(i))) {
        return true;
      }
    }
    return false;
  }

  private static boolean isMatchHere(String r, String s) {
    // Case (1.)
    if (r.isEmpty()) {
      return true;
    }

    // Case (2) : ends with '$'.
    if ("$".equals(r)) {
      return s.isEmpty();
    }

    // Case (4.)
    if (r.length() >= 2 && r.charAt(1) == '*') {
      for (int i = 0; i < s.length()
          && (r.charAt(0) == '.' || r.charAt(0) == s.charAt(i)); ++i) {
        if (isMatchHere(r.substring(2), s.substring(i + 1))) {
          return true;
        }
      }
      return isMatchHere(r.substring(2), s);
    }

    // Case (3.)
    return !s.isEmpty() && (r.charAt(0) == '.' || r.charAt(0) == s.charAt(0))
        && isMatchHere(r.substring(1), s.substring(1));
  }
  // @exclude

  public static void main(String[] args) {
    assert (isMatch(".", "a"));
    assert (isMatch("a", "a"));
    assert (!isMatch("a", "b"));
    assert (isMatch("a.9", "aW9"));
    assert (!isMatch("a.9", "aW19"));
    assert (isMatch("^a.9", "aW9"));
    assert (!isMatch("^a.9", "baW19"));
    assert (isMatch(".*", "a"));
    assert (isMatch(".*", ""));
    assert (isMatch("c*", "c"));
    assert (!isMatch("aa*", "c"));
    assert (isMatch("ca*", "c"));
    assert (isMatch(".*", "asdsdsa"));
    assert (isMatch("9$", "xxxxW19"));

    assert (isMatch(".*a", "ba"));

    assert (isMatch(".*a", "ba"));
    assert (isMatch("^a.*9$", "axxxxW19"));

    assert (isMatch("^a.*W19$", "axxxxW19"));
    assert (isMatch(".*a.*W19", "axxxxW19123"));
    assert (!isMatch(".*b.*W19", "axxxxW19123"));
    assert (isMatch("n.*", "n"));
    assert (isMatch("a*n.*", "an"));
    assert (isMatch("a*n.*", "ana"));
    assert (isMatch("a*n.*W19", "anaxxxxW19123"));
    assert (isMatch(".*a*n.*W19", "asdaaadnanaxxxxW19123"));
    assert (isMatch(".*.*.a*n.*W19", "asdaaadnanaxxxxW19123"));
  }
}
